bootstore/src/schemes/v0/state.rs

```diff
@@ -0,0 +1,594 @@
+    pub fn next_peer(&self, current: &Baseboard) -> Option<Baseboard> {
+        if let Some(index) = self.peers.iter().position(|x| x == current) {
+            let next_index = (index + 1) % self.peers.len();
+            self.peers.iter().nth(next_index).cloned()
+        } else {
+            self.peers.first().cloned()
+        }
+    }
+
+    // Send a `GetShare` request to the newly connected peer if necessary
+    pub fn on_connect(&self, peer: Baseboard, rack_uuid: Uuid) -> Output {
+        if let RackSecretState::Retrieving { request_id, from, .. } =
+            &self.rack_secret_state
+        {
+            // We don't have a share from the peer and we weren't previously
+            // connected to the peer.
+            if !from.contains(&peer) && !self.peers.contains(&peer) {
+                let request = Request {
+                    id: *request_id,
+                    type_: RequestType::GetShare { rack_uuid },
+                };
+                return Output {
+                    persist: false,
+                    envelopes: vec![Envelope { to: peer, msg: request.into() }],
+                    api_output: None,
+                };
+            }
+        }
+        Output::none()
+    }
+}
+
+/// Metadata associated with a request that is keyed by `Baseboard`
+#[derive(Debug)]
+pub struct RequestMetadata {
+    pub request_id: Uuid,
+    pub expiry: Ticks,
+}
+
+#[derive(Zeroize, ZeroizeOnDrop)]
+pub struct Shares(Vec<Vec<u8>>);
+
+impl Shares {
+    fn insert(&mut self, share: Vec<u8>) {
+        self.0.push(share);
+    }
+}
+
+impl Debug for Shares {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Shares").finish()
+    }
+}
+
+/// The state of rack secret share retrieval
+///
+/// Primarily this is whether we are in the process of retrieving enough shares to
+/// recompute the rack secret or if we have already recomputed it.
+#[derive(Debug)]
+pub enum RackSecretState {
+    /// We are not currently retrieving any shares and we don't have a computed
+    /// rack secret
+    Empty,
+
+    /// We only store the shares in memory until some expiry deadline.
+    /// However, we extend this deadline each time a share is learned to enable
+    /// computation of the rack secret. If no shares are added for a while, and
+    /// we still don't have enough shares to compute the rack secret then we
+    /// drop all the shares.
+    Retrieving {
+        // We use a separate BTreeSet to track who we retrieved shares form so
+        // that we can wrap the shares with a `Secret`
+        request_id: Uuid,
+        from: BTreeSet<Baseboard>,
+        shares: Shares,
+        expiry: Ticks,
+        threshold: usize,
+        share_digests: Vec<Sha3_256Digest>,
+    },
+    /// We only store the computed secret in memory for so long after it has
+    /// been computed
+    Computed { secret: RackSecret, expiry: Ticks },
+}
+
+impl RackSecretState {
+    pub fn load(
```

Comment:
The fact that this is a separate state machine seems odd to me, and leads to weird semantics.

For example, what if I call `load(...)` on a `RackSecretState` that _already_ has a rack secret, but I pass in a different `Uuid` / `Baseboard` / local share / etc?  That's totally possible, given the API, but the current behavior is to just ignore that info and return the previous secret.

It may be more intuitive to put these transitions into the main `State` object, instead of having dueling (unsynchronized) state machines.

Reply:
No longer relevant as of [98ada27](https://github.com/oxidecomputer/omicron/pull/3385/commits/98ada27b53d4429463cde94cc91bedec0bc4ef98)